docs: add App Runner-WAF addon Documentation (#4464)
aws/copilot-cli: site/content/blogs/apprunner-waf.en.md

# Using WAF With App Runner in Copilot

Posted On: Feb 08, 2023

**Siddharth Vohra**, Software Development Engineer, AWS App Runner

You can now associate your AWS Web Application Firewall (WAF) Web access control lists (web ACLs) with your App Runner service - all in Copilot!

[AWS Web Application Firewall (WAF)](https://docs.aws.amazon.com/waf/latest/developerguide/waf-chapter.html) helps you monitor the HTTP(S) requests that are forwarded to your web applications,
and allows you to control access to your content. 
You can protect your application by blocking or allowing web requests based on criteria that you specify, 
such as the IP addresses that requests originate from or the values of query strings.

> Today, AWS announced the WAF support for AWS App Runner. 

Reviewer: An idea: if there is a blog post that AppRunner + WAF is making, we can add the link to that blog post here as well!
```suggestion
Today, AWS announced the [WAF support for AWS App Runner](/link/to/post). 
```

Author: Looks like we'll only be able to do this with a new pull request early next week.

      Name:
        Type: String
        Description: The name of the service, job, or workflow being deployed.
      ServiceARN:
        Type: String
        Default: ""
        Description: The ARN of the service being deployed.
    
    Resources:
      # Configuration of the WAF Web ACL you want to asscoiate with 
      # your App Runner service.
      Firewall:
        Metadata:
          'aws:copilot:description': 'Associating your App Runner service with your WAF WebACL'
        Type: AWS::WAFv2::WebACLAssociation
        Properties: 
          ResourceArn: !Sub ${ServiceARN}
          WebACLArn:  <paste your WAF Web ACL ARN here> # Paste your WAF Web ACL ARN here.
    ```

=== "addons.parameters.yml"
      ```yaml
      Parameters:
        ServiceARN: !Ref Service
      ```

### Step 4: Populate your Web ACL ARN in `waf.yml`

Open `waf.yml` and replace `<paste your WAF Web ACL ARN here>` with your the real ARN of your Web ACL resource. For example:   
```yaml
Resources:
  # Configuration of the WAF Web ACL you want to associate with 
  # your App Runner service
  Firewall:
    Metadata:

>       'aws:copilot:description': 'Associating your App Runner service with your WAF WebACL'

Reviewer 2: ```suggestion
      'aws:copilot:description': 'Associating your App Runner service with your WAF Web ACL'
```

Author: Done!

    Type: AWS::WAFv2::WebACLAssociation
    Properties: 
      ResourceArn: !Sub ${ServiceARN}
      WebACLArn: arn:aws:wafv2:us-east-2:123456789138:regional/webacl/mytestwebacl/3df43564-be9f-47ce-a12b-3a577d2d8913
```
 

### Step 5: Deploy your service 

> Finally, run `copilot svc deploy`! Your Request-Driven Web service will now be deployed along with a WAF WebACL!  

Reviewer 2: ```suggestion
Finally, run `copilot svc deploy`! Your Request-Driven Web service will now be deployed along with a WAF Web ACL!  
```

Author: Done!

          ├── manifest.yml
          └── addons/
              ├── waf.yml 
              └── addons.parameters.yml
  ```

Copy and paste the following content to the respective files:  

=== "waf.yml"
```yaml
#Addon template to add WAF configuration to your App Runner service.
  
Parameters:
  App:
    Type: String
    Description: Your application's name.
  Env:
    Type: String
    Description: The environment name your service, job, or workflow is being deployed to.
  Name:
    Type: String
    Description: The name of the service, job, or workflow being deployed.
  ServiceARN:
    Type: String
    Default: ""
    Description: The ARN of the service being deployed.
  
Resources:
  # Configuration of the WAF Web ACL you want to asscoiate with 
  # your App Runner service.
  Firewall:
    Metadata:
      'aws:copilot:description': 'Associating your App Runner service with your WAF Web ACL'
    Type: AWS::WAFv2::WebACLAssociation
    Properties: 
      ResourceArn: !Sub ${ServiceARN}
      WebACLArn:  <paste your WAF Web ACL ARN here> # Paste your WAF Web ACL ARN here.
```

=== "addons.parameters.yml"  
```yaml
Parameters:
  ServiceARN: !Ref Service

> ```

Reviewer: Tabs here are actually necessary for the code snippet to be rendered under the tab  "addons.parameters.yml".

Can we restore the tabs in my commit here https://github.com/aws/copilot-cli/pull/4464/commits/c8a545c3b24a2598095a7db1e54a2931630b29b4 ? I made sure they render properly!

Author: Oh okay. Apologies. Restoring the tabs.

### Step 4: Populate your Web ACL ARN in `waf.yml`

Open `waf.yml` and replace `<paste your WAF Web ACL ARN here>` with your the real ARN of your Web ACL resource. For example:   

>    ```yaml

Reviewer: Why do we need the extra `tab` here 🤔 ? Does it render nicely when you run `make site-local`?

Author: Looks like this was a typo as well. Fixed it.

> # Using WAF With AppRunner in Copilot

Reviewer: I see we sometimes use `AppRunner` and sometimes `App Runner`. Which one is more correct 🤔 ? We can pick one and be consistent across this blog!

Author: Since our official name is App Runner - that should be the correct one! I'll fix this, apologies.

Author: Done

Posted On: Feb 08, 2023

**Siddharth Vohra**, Software Development Engineer, AWS App Runner

You can now associate your AWS WAF WebACLs with your App Runner service - all in Copilot!

> AWS Web Application Firewall (WAF) helps you monitor requests that are forwarded to your web applications and allows you to control access to your content. To do this, you need to create WAF Web Application Control List (Web ACL) which contains many rule options for your application. You can add multiple rules and create a Web ACL. Once you’ve created a Web ACL through AWS WAF Console or the AWS CLI, copy and save the Web ACL ARN. Now follow the steps below.

Reviewer 2: ```suggestion
AWS Web Application Firewall (WAF) helps you monitor requests that are forwarded to your web applications and allows you to control access to your content. To do this, you need to create a WAF Web Application Control List (ACL) which contains many rule options for your application. You can add multiple rules and create a Web ACL. Once you’ve created a Web ACL through AWS WAF Console or the AWS CLI, copy and save its ARN. To use the WAF ACL with your App Runner service managed by Copilot, follow the steps below.
```

> AWS Web Application Firewall (WAF) helps you monitor requests that are forwarded to your web applications and allows you to control access to your content. To do this, you need to create WAF Web Application Control List (Web ACL) which contains many rule options for your application. You can add multiple rules and create a Web ACL. Once you’ve created a Web ACL through AWS WAF Console or the AWS CLI, copy and save the Web ACL ARN. Now follow the steps below.

Reviewer: We can maybe mention this page: https://docs.aws.amazon.com/waf/latest/developerguide/web-acl-creating.html. Something along the line:
```md
To do this, you need to create a WAF Web Application Control List (ACL) which can contain many rule 
options for your application. See [here](https://docs.aws.amazon.com/waf/latest/developerguide/web-acl-creating.html) 
for more on creating your own Web ACL. 
```

Author: Yes! Extremely helpful. Done.

>  Step 1: If you don’t already have an App Runner service created/deployed, run `copilot init` to create and configure an App Runner service. (Do not deploy it just yet!)  

Reviewer 2: Maybe suggest them to use `svc init`?
```suggestion
 Step 1: If you don’t have an App Runner service created/deployed, run `copilot svc init` to create and configure an App Runner service.
```

Reviewer: Just to add on PH's suggestion - It's okay to have deployed a service, and add an addon to it afterwards! Hence is`(Do not deploy it just yet!)`  is a tiny bit misleading

Author: Understood, yes

>  Step 1: If you don’t already have an App Runner service created/deployed, run `copilot init` to create and configure an App Runner service. (Do not deploy it just yet!)  

Reviewer: We might want to mention "Request-Driven Web Service" here:
```suggestion
 Step 1: If you don’t already have an App Runner service created/deployed, run `copilot svc init --type "Request-Driven Web Service"` to create and configure an App Runner service.
```

Author: Yes! Done

> Step 2: Go to the service's directory (users/\<your username\>/copilot/\<your App Runner service name\>). If you don't have a folder called 'addons' in this directory already, create a new folder and name it 'addons'. Now, go to the AR-WAF Addon discussion (Hyperlink coming soon) and follow the steps.  Your folders would now look like this:  

Reviewer 2: Why do we need them to go to "AR-WAF Addon discussion"? Isn't this blog post specific enough?
```suggestion
Step 2: Go to the service's directory (users/\<your username\>/copilot/\<your App Runner service name\>). If you don't have a folder called 'addons' in this directory, create a new 'addons' folder. Now, go to the AR-WAF Addon discussion (Hyperlink coming soon) and follow the steps.  Your folders would now look like this:  
```

Author: This blog post is indeed quiet specific but the show-and-tell section is where other addons sit too and it might be a good area to increase visibility. Also, we could also then get direct feedback from users regarding the addon.

> Step 2: Go to the service's directory (users/\<your username\>/copilot/\<your App Runner service name\>). If you don't have a folder called 'addons' in this directory already, create a new folder and name it 'addons'. Now, go to the AR-WAF Addon discussion (Hyperlink coming soon) and follow the steps.  Your folders would now look like this:  

Reviewer: I agree that we don't need to drop the discussion link here. This blog post is great enough for users to follow - we don't want to break the flow.

We can add a line at the end of this blog post, but not in the middle:
```
We posted these step in our [GitHub "Show and tell" discussion section](link) as well! If you have any questions, feedbacks or requests that are related to AppRunner's WAF support, feel free to drop a comment there!
```

Author: That makes sense. Will update this.

Author: Done

  ```term
  .
  └── copilot
      └── <your App Runner service name>
          └── addons

>               └── ar_waf_addon.yml 

Reviewer: nit: what do you think of just `waf.yml`?

Author: We could do that but we also want to make sure that users remember that this addon is only for AR services and may not work with other services supported by Copilot. Do let me know though, I don't mind changing it at all. If you do want it changed to waf.yml, I'll do it.

Author: Done!

              └── addons.parameters.yml
  ```

> and `ar_waf-addon.yml` would look like  

Reviewer: Given that we don't direct readers to show-and-tell in the middle of the blog post, we can reword this sentence to something like "copy and paste the following CloudFormation template into waf.yml"

Author: Done!

  ```yaml
#Addon template to add WAF configuration to your App Runner service.

Parameters:
  App:
    Type: String
    Description: Your application's name.
  Env:
    Type: String
    Description: The environment name your service, job, or workflow is being deployed to.
  Name:
    Type: String
    Description: The name of the service, job, or workflow being deployed.
  ServiceARN:
    Type: String
    Default: ""
    Description: The ARN of the service being deployed.

Resources:
  # Configuration of the WAF Web ACL you want to asscoiate with 
  # your App Runner service
  Firewall:
    Metadata:
      'aws:copilot:description': 'Associating your App Runner service with your WAF WebACL'
    Type: AWS::WAFv2::WebACLAssociation
    Properties: 
      ResourceArn: !Sub ${ServiceARN}

>       WebACLArn:  <paste your WAF Web ACL ARN here> #Paste your WAF Web ACLL ARN here

Reviewer: Pretty nice to have these comments!!

Author: :)

  ```

while `addons.parameters.yml` would look like:  
  ```yaml
  Parameters:
    ServiceARN: !Ref Service
  ```

Step 3: Open `ar_waf_addon.yml` and edit it to add your WebACL ARN where it is required - `Firewall.Properties.WebACLArn`. For example:   
  ```yaml
Resources:
  # Configuration of the WAF Web ACL you want to asscoiate with 
  # your App Runner service
  Firewall:
    Metadata:
      'aws:copilot:description': 'Associating your App Runner service with your WAF WebACL'
    Type: AWS::WAFv2::WebACLAssociation
    Properties: 
      ResourceArn: !Sub ${ServiceARN}
      WebACLArn: arn:aws:wafv2:us-east-2:637867102138:regional/webacl/mytestwebacl/3df32464-be9f-47ce-a12b-3a466c1c8913
  ```
 

Step 4: Now save the file and run `copilot svc deploy` to deploy your new App Runner service. Your App Runner service will now be deployed along with a WAF WebACL!  

> Some considerations:  

Reviewer 2: We could probably use https://github.com/aws/copilot-cli/blob/60cf1792200c86a6c317d846e57ddc08a9286af1/site/content/blogs/release-v125.en.md?plain=1#L23

Author: Done!

# Using WAF With App Runner in Copilot

Posted On: Feb 08, 2023

**Siddharth Vohra**, Software Development Engineer, AWS App Runner

You can now associate your AWS WAF WebACLs with your App Runner service - all in Copilot!

We posted these step in our [GitHub "Show and tell" discussion section](LINK TO BE ADDED) as well! If you have any questions, feedbacks or requests that are related to AppRunner's WAF support, feel free to drop a comment there!

AWS Web Application Firewall (WAF) helps you monitor requests that are forwarded to your web applications and allows you to control access to your content.To do this, you need to create a WAF Web Application Control List (ACL) which can contain many rule 

> options for your application. See [here](https://docs.aws.amazon.com/waf/latest/developerguide/web-acl-creating.html) for more on creating your own Web ACL. Once you’ve created a Web ACL through AWS WAF Console or the AWS CLI, copy and save its ARN. To use the WAF ACL with your App Runner service managed by Copilot, follow the steps below.  

Reviewer 2: ```suggestion
AWS Web Application Firewall (WAF) helps you monitor requests that are forwarded to your web applications and allows you to control access to your content. To do this, you need to create a WAF Web Application Control List (ACL) with rule options for your application. See [here](https://docs.aws.amazon.com/waf/latest/developerguide/web-acl-creating.html) for more on creating your own Web ACL. Once you’ve created a Web ACL through AWS WAF Console or the AWS CLI, copy and save its ARN. To use the WAF ACL with your App Runner service managed by Copilot, follow the steps below.  
```

Reviewer: the intro is reworded largely - I'm resolving this convo! if you have other stuff to add on top of the current version feel free to unresolveeeee

Step 1: If you don’t have an App Runner service created/deployed, run `copilot svc init --type "Request-Driven Web Service"` to create and configure an App Runner service.

> Step 2: Go to the service's directory (users/\<your username\>/copilot/\<your App Runner service name\>). If you don't have a folder called 'addons' in this directory, create a new 'addons' folder. In the addons folder, create two new files: `ar_waf-addon.yml` and `addons.parameters.yml`. Your folders would now look like this:  

Reviewer 2: ```suggestion
Step 2: Go to the service's directory (users/\<your username\>/copilot/\<your App Runner service name\>). If you don't have a folder called 'addons' in this directory, create a new 'addons' folder. In the addons folder, create two new files: `ar_waf-addon.yml` and `addons.parameters.yml`.
```

Author: Done by @Lou1415926

>   ```term

Reviewer 2: Can we do something like [this](https://github.com/aws/copilot-cli/blob/212611525869a81396c6f1b57a479cc1ea5317be/site/content/docs/manifest/lb-web-service.en.md?plain=1#L3) so that these code snippets can be combined with different titles? Same for the `waf.yaml` file below.

Reviewer: this is done ✔️

  .
  └── copilot
      └── <your App Runner service name>
          └── addons
              └── waf.yml 
              └── addons.parameters.yml
  ```

> Copy, paste and save the content below so that your `waf.yml` looks like:

Reviewer 2: We can get rid of those "copy, paste and save..." for step 2.

Author: Fixed by @Lou1415926 as well.
